icu4j/main/classes/core/src/com/ibm/icu/text/DateIntervalFormat.java

```diff
@@ -755,76 +971,83 @@ public final synchronized StringBuffer format(Calendar fromCalendar,
                  * the smallest calendar field in pattern,
                  * return single date format.
                  */
-                return fDateFormat.format(fromCalendar, appendTo, pos);
+                return fDateFormat.format(fromCalendar, appendTo, pos, attributes);
             }
 
-            return fallbackFormat(fromCalendar, toCalendar, fromToOnSameDay, appendTo, pos);
+            return fallbackFormat(fromCalendar, toCalendar, fromToOnSameDay, appendTo, pos,
+                    output, attributes);
         }
 
-        // If the first part in interval pattern is empty, 
+        // If the first part in interval pattern is empty,
         // the 2nd part of it saves the full-pattern used in fall-back.
         // For a 'real' interval pattern, the first part will never be empty.
         if ( intervalPattern.getFirstPart() == null ) {
             // fall back
             return fallbackFormat(fromCalendar, toCalendar, fromToOnSameDay, appendTo, pos,
-                                    intervalPattern.getSecondPart());
+                    output, attributes, intervalPattern.getSecondPart());
         }
         Calendar firstCal;
         Calendar secondCal;
         if ( intervalPattern.firstDateInPtnIsLaterDate() ) {
+            if (output != null) {
+                output.register(1);
+            }
             firstCal = toCalendar;
             secondCal = fromCalendar;
         } else {
+            if (output != null) {
+                output.register(0);
+            }
             firstCal = fromCalendar;
             secondCal = toCalendar;
         }
         // break the interval pattern into 2 parts
-        // first part should not be empty, 
+        // first part should not be empty,
         String originalPattern = fDateFormat.toPattern();
         fDateFormat.applyPattern(intervalPattern.getFirstPart());
-        fDateFormat.format(firstCal, appendTo, pos);
+        fDateFormat.format(firstCal, appendTo, pos, attributes);
+        // Only accept the first instance of the field
+        if (pos.getEndIndex() > 0) {
+            pos = new FieldPosition(0);
+        }
         if ( intervalPattern.getSecondPart() != null ) {
             fDateFormat.applyPattern(intervalPattern.getSecondPart());
-            FieldPosition otherPos = new FieldPosition(pos.getField());
-            fDateFormat.format(secondCal, appendTo, otherPos);
-            if (pos.getEndIndex() == 0 && otherPos.getEndIndex() > 0) {
-                pos.setBeginIndex(otherPos.getBeginIndex());
-                pos.setEndIndex(otherPos.getEndIndex());
-            }
+            fDateFormat.format(secondCal, appendTo, pos, attributes);
```

Comment:
Maybe I'm missing something, but aren't you overriding the value that it's stored in `pos`?
Or will SimpleDateFormat avoid overriding `pos` if `getEndIndex() > 0`?

Reply:
The style I'm using in Java is to unset the reference to the user's `pos` once it is populated.  I believe the first time `pos` could change is in the first call to `fDateFormat.format` on line 1008.  On line 1010, I check if `pos` was updated, and if it was, I set it to a dummy instance.  There were three places I had to do this hack, all with the same comment `// Only accept the first instance of the field`.  Please verify that I caught everywhere I needed to make this change.

In C++, of course, we have FieldPositionHandler, which is a much nicer abstraction to use.

Reply:
I see, looks good.